composeml/label_maker.py

```diff
@@ -4,141 +4,333 @@
+        offset = value
+
+    return offset
 
 
 class LabelMaker:
     """Automatically makes labels for prediction problems."""
 
-    def __init__(self, target_entity, time_index, labeling_function, window_size):
-        """
-        Creates an instance of label maker.
+    def __init__(self, target_entity, time_index, labeling_function, window_size=None):
+        """Creates an instance of label maker.
 
         Args:
             target_entity (str) : Entity on which to make labels.
             time_index (str): Name of time column in the data frame.
             labeling_function (function) : Function that transforms a data slice to a label.
             window_size (str or int) : Duration of each data slice.
+                The default value for window size is all future data.
         """
-        assert_valid_offset(window_size)
-
         self.target_entity = target_entity
         self.time_index = time_index
         self.labeling_function = labeling_function
         self.window_size = window_size
 
-    def _preprocess(self, df):
-        if df.index.name != self.time_index:
-            df = df.set_index(self.time_index)
+        if self.window_size is not None:
+            self.window_size = to_offset(self.window_size)
 
-        if 'time' not in str(df.index.dtype):
-            df.index = df.index.astype('datetime64[ns]')
+    def get_slices(self, df, gap=None, min_data=None, drop_empty=True):
+        """Generate data slices.
 
-        return df
+        Args:
+            df (DataFrame) : Data frame to generate data slices.
+            gap (str) : Time between slices. Default value is window size.
+            min_data (int or str or Timestamp) : Threshold to cutoff data.
+            drop_empty (bool) : Whether to drop empty slices.
+
+        Returns:
+            DataFrame, dict : Returns a data slice and metadata about the data slice.
+        """
+        self.window_size = self.window_size or len(df)
+        gap = to_offset(gap or self.window_size)
+
+        df = df.loc[df.index.notnull()]
+        df.sort_index(inplace=True)
+
+        if df.empty:
+            return
+
+        threshold = min_data or df.index[0]
+        df, cutoff_time = cutoff_data(df=df, threshold=threshold)
+
+        if df.empty:
+            return
+
+        if isinstance(gap, int):
+            cutoff_time = df.index[0]
+
+        metadata = {'slice': 0, 'min_data': cutoff_time}
+
+        def iloc(index, i):
+            if i < index.size:
+                return index[i]
+
+        while not df.empty and cutoff_time <= df.index[-1]:
+            if isinstance(self.window_size, int):
+                df_slice = df.iloc[:self.window_size]
+                window_end = iloc(df.index, self.window_size)
+
+            else:
+                window_end = cutoff_time + self.window_size
+                df_slice = df[:window_end]
+
+                if not df_slice.empty:
+                    # exclude last row to avoid overlap
```

Comment:
having a hard time understanding why we need to can the overlap, can you explain further?

Reply:
Related to [this discussion](https://github.com/FeatureLabs/compose/pull/44#discussion_r312073184). Essentially, pandas includes both endpoints when slicing by time. This results in one endpoint overlapping in consecutive data slices. Discussed in [this section of pandas documentation](https://pandas.pydata.org/pandas-docs/version/0.19/gotchas.html#endpoints-are-inclusive).

Reply:
added comment with link to pandas docs